File: practice-python/bfs/rotten_orange.py

from typing import List
from collections import deque

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        empty = 0
        fresh = 1
        rotten = 2

        max_row = len(grid) - 1
        max_col = len(grid[0]) - 1
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        queue = deque()
        visited = set()

        num_fresh_orange = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == fresh:
                    num_fresh_orange += 1
                elif grid[i][j] == rotten:
                    queue.append((i, j))
                    visited.add((i, j))

        def get_neighbor(r, c):
            for dr, dc in directions:
                nr, nc = r + dr, c + dc

                if not (0 <= nr <= max_row and 0 <= nc <= max_col):
                    continue
                if grid[nr][nc] == empty:
                    continue

                yield (nr, nc)

        time_elapsed = 1
        while queue:
            row, col = queue.popleft()
            if len(queue) == 0:
                time_elapsed += 1

            if grid[row][col] == fresh:
                grid[row][col] = rotten
                num_fresh_orange -= 1

            for neigobor in get_neighbor(row, col):
                if neigobor in visited:
                    continue
                queue.append(neigobor)
                visited.add(neigobor)

        return time_elapsed if num_fresh_orange == 0 else -1

sol = Solution()

grid = [[2,1,1],[1,1,0],[0,1,1]]
ans = 4
res = sol.orangesRotting(grid)
if res == ans:
    print("Pass")
else:
    print("Failed", res)

File: practice-cpp/bfs/rotten_orange.cpp

#include <deque>
#include <vector>
#include <iostream>
#include <utility>

using std::vector;
using std::deque;
using std::pair;
using std::make_pair;
using std::cout;


class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        
        int n = grid.size();
        int m = grid[0].size();
        int max_row = n - 1, max_col = m - 1;
        deque<pair<int, int>> queue;
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        vector<pair<int, int>> directions{
          {-1, 0}, {1, 0}, {0, 1}, {0, -1}};

        int total_fresh_orange = 0;
        // init queue
        for(int i =0; i < n; i++) {
          for (int j = 0; j < m; j++) {
            if (grid[i][j] == 2) {
              queue.push_back(make_pair(i, j));
            } else if (grid[i][j] == 1) {
              total_fresh_orange ++;
            }
          }
        }

        if (total_fresh_orange == 0) {
          return 0;
        }

        int time_elapsed = -1;
        while (!queue.empty()) {
          time_elapsed++;
          int num_fresh = queue.size();

          while (num_fresh > 0) {
            num_fresh--;
            int row = queue.front().first;
            int col = queue.front().second;
            queue.pop_front();

            for (auto [dr, dc]: directions) {
              int new_row = row + dr;
              int new_col = col + dc;

              // skip invalid
              if (!(new_row >=0 and new_row <= max_row and new_col >= 0 and new_col <= max_col)) {
                continue;
              }

              if (grid[new_row][new_col] == 1 and !visited[new_row][new_col]) {
                queue.push_back(make_pair(new_row, new_col));
                visited[new_row][new_col] = true;
                total_fresh_orange--;
              }
            }
          }
        }

        if (total_fresh_orange == 0) {
          return time_elapsed;
        } 
        return -1;
    }
};

int main() {
  Solution sol;

  vector<vector<int>> grid{
    {2, 1, 1},
    {1, 1, 0},
    {0, 1, 1}
  };
  int ans = 4;

  auto res = sol.orangesRotting(grid);
  if (res == ans) {
    cout << "Passed\n";
  } else {
    cout << "Failed, " << res << "\n";
  }

  return 0;
}